File: backend/app/domain/services/evaluation.py
import uuid
from typing import Any

from app.core.exceptions import NotFoundError
from app.core.redis import get_cached_asset_scores
from app.domain.services.base import BaseService
from app.infrastructure.repositories.asset_scores import AssetScoresRepository


class EvaluationService(BaseService):
    def __init__(self, repo: AssetScoresRepository):
        self.repo = repo

    def get_asset_scores(self, asset_id: uuid.UUID, model_version: str = "v1.0.0") -> dict[str, Any]:
        cached = get_cached_asset_scores(str(asset_id))
        if cached and cached.get("model_version") == model_version:
            return cached

        scores = self.repo.get(asset_id, model_version)
        if not scores:
            raise NotFoundError("Asset scores not found")

        return {
            "asset_id": str(scores.asset_id),
            "model_version": scores.model_version,
            "recommendation_score": float(scores.recommendation_score),
            "quality_score": float(scores.quality_score),
            "valuation_score": float(scores.valuation_score),
            "generated_at": scores.generated_at
        }

File: backend/src/routes/evaluation/evaluation.ts
import { Router } from "express";
import { prisma } from "../../prisma";
import { NotFoundError } from "../../lib/errors";
import { requireUuidParam } from "../../lib/validation";
import { getCachedAssetScores } from "../../lib/evaluation/cache";

export const evaluationRouter = Router();

interface AssetScoresResponse {
  asset_id: string;
  model_version: string;
  recommendation_score: number | null;
  quality_score: number | null;
  valuation_score: number | null;
  unavailable_inputs: unknown;
  generated_at: string | null;
}

// Port of EvaluationService.get_asset_scores.
evaluationRouter.get("/assets/:assetId/scores", async (req, res) => {
  const { assetId } = req.params;
  requireUuidParam(assetId, "asset_id");
  const modelVersion = typeof req.query.model_version === "string" ? req.query.model_version : "v1.0.0";

  const cached = await getCachedAssetScores<AssetScoresResponse>(assetId);
  if (cached && cached.model_version === modelVersion) {
    res.json(cached);
    return;
  }

  const scores = await prisma.assetScore.findUnique({
    where: { assetId_modelVersion: { assetId, modelVersion } },
  });
  if (!scores) throw new NotFoundError("Asset scores not found");

  res.json({
    asset_id: scores.assetId,
    model_version: scores.modelVersion,
    recommendation_score: scores.recommendationScore != null ? Number(scores.recommendationScore) : null,
    quality_score: scores.qualityScore != null ? Number(scores.qualityScore) : null,
    valuation_score: scores.valuationScore != null ? Number(scores.valuationScore) : null,
    unavailable_inputs: scores.unavailableInputs ?? [],
    generated_at: scores.generatedAt ? scores.generatedAt.toISOString() : null,
  });
});
